File: pkg/far/faregex.go
package far

import "regexp"

// Faregex is a finder and replacer base on regexp.
type Faregex struct {
	pattern  *regexp.Regexp
	template string
}

func (f *Faregex) See(text string) (diff Diffs) {

	if f.pattern != nil {

		i := 0
		for _, m := range f.pattern.FindAllStringSubmatchIndex(text, -1) {
			if len(m) < 2 {
				continue // should never reach
			}

			ret := text[i:m[0]]
			del := text[m[0]:m[1]]
			ins := string(f.pattern.ExpandString(nil, f.template, text, m))

			if del == ins {
				continue
			}
			if len(ret) != 0 {
				diff = append(diff, Diff{Type: DiffRetain, Text: ret})
			}
			if len(del) != 0 {
				diff = append(diff, Diff{Type: DiffDelete, Text: del})
			}
			if len(ins) != 0 {
				diff = append(diff, Diff{Type: DiffInsert, Text: ins})
			}

			i = m[1]
		}

		if i != len(text) {
			diff = append(diff, Diff{Type: DiffRetain, Text: text[i:]})
		}
	}

	if diff == nil {
		diff = Diffs{{Type: DiffRetain, Text: text}}
	}

	return diff
}

func (f *Faregex) Empty() bool {
	return f.pattern == nil
}

func (f *Faregex) Pattern() string {
	if f.pattern != nil {
		return f.pattern.String()
	}
	return ""
}

func (f *Faregex) Template() string {
	return f.template
}

func (f *Faregex) SetPattern(text string) (err error) {
	var pat *regexp.Regexp
	if text == "" {
		pat = nil
	} else {
		pat, err = regexp.Compile(text)
	}

	if err == nil {
		f.pattern = pat
	}
	return
}

func (f *Faregex) SetTemplate(text string) (err error) {
	f.template = text
	return
}

File: src/far/faregex.rs
//! This file is based on:
//! https://github.com/wiryls/far/blob/46271a4a1dfb69c321aea4564337b838132b775f/pkg/far/faregex.go
use regex::{Regex, Error};
use super::diff::{Diff, Patch};

#[derive(Clone, Debug)]
pub struct Faregex {
    pattern  : Option<Regex>,
    template : String,
}

/// An find-and-replace tool base on regexp.
impl Faregex {

    /// Create a Faregex.
    pub const fn new() -> Self {
        Faregex {
            pattern : None,
            template: String::new(),
        }
    }

    /// See the difference.
    pub fn see(&self, text : &str) -> Diff {
        match &self.pattern {
            None => Diff::default(),
            Some(regexp) => {
                let mut last : usize = 0;
                let mut diff = Vec::new();
                for cap in regexp.captures_iter(text) {
                    // Why is the Item of SubCaptureMatches.Iterator an
                    // Option<Match> ???
                    // https://docs.rs/regex/1.5.4/regex/struct.SubCaptureMatches.html#associatedtype.Item
                    let m = match cap.get(0) {
                        None    => continue,
                        Some(m) => m,
                    };

                    let     ret = &text[last..m.start()];
                    let     del = &text[m.range()];
                    let mut ins = String::new();
                    cap.expand(self.template.as_str(), &mut ins);

                    if del != ins {
                        if !ret.is_empty() {
                            diff.push(Patch::Retain(last..m.start()));
                        }
                        if !del.is_empty() {
                            diff.push(Patch::Delete(m.range()));
                        }
                        if !ins.is_empty() {
                            diff.push(Patch::Insert(ins));
                        }
                        last = m.end();
                    }
                }
                if 0 != last && last != text.len() {
                    diff.push(Patch::Retain(last..text.len()));
                }
                Diff::from_vec(diff)
            },
        }
    }

    /// is it always returns `Diff::Retain(String)`
    pub fn is_empty(&self) -> bool {
        self.pattern.is_none()
    }

    /// set a pattern for regex.
    pub fn set_pattern(&mut self, pattern : &str) -> Option<Error> {
        match Regex::new(pattern) {
            Ok (r) => {self.pattern = Some(r); None},
            Err(e) => Some(e),
        }
    }

    /// read the pattern for regex.
    pub fn get_pattern(&self) -> &str {
        match &self.pattern {
            None    => &"",
            Some(p) => p.as_str(),
        }
    }

    /// set a template for replacing.
    pub fn set_template(&mut self, template : &str) {
        self.template = template.to_owned();
    }

    /// get the template for replacing.
    pub fn get_template(&self) -> &str {
        &self.template[..]
    }
}
